storage/authorization.go

```diff
@@ -0,0 +1,220 @@
+
+type authentication string
+
+const (
+	sharedKey             authentication = "sharedKey"
+	sharedKeyForTable     authentication = "sharedKeyTable"
+	sharedKeyLite         authentication = "sharedKeyLite"
+	sharedKeyLiteForTable authentication = "sharedKeyLiteTable"
+
+	// headers
+	headerAuthorization     = "Authorization"
+	headerContentLength     = "Content-Length"
+	headerDate              = "Date"
+	headerXmsDate           = "x-ms-date"
+	headerXmsVersion        = "x-ms-version"
+	headerContentEncoding   = "Content-Encoding"
+	headerContentLanguage   = "Content-Language"
+	headerContentType       = "Content-Type"
+	headerContentMD5        = "Content-MD5"
+	headerIfModifiedSince   = "If-Modified-Since"
+	headerIfMatch           = "If-Match"
+	headerIfNoneMatch       = "If-None-Match"
+	headerIfUnmodifiedSince = "If-Unmodified-Since"
+	headerRange             = "Range"
+)
+
+func (c Client) addAuthorizationHeader(verb, url string, headers *map[string]string, auth authentication) error {
+	authHeader, err := c.getSharedKey(verb, url, *headers, auth)
+	if err != nil {
+		return err
+	}
+	(*headers)[headerAuthorization] = authHeader
+	return nil
+}
+
+func (c Client) getSharedKey(verb, url string, headers map[string]string, auth authentication) (string, error) {
+	canRes, err := c.buildCanonicalizedResource(url, auth)
+	if err != nil {
+		return "", err
+	}
+
+	canString := c.buildCanonicalizedString(verb, headers, canRes, auth)
+	return c.createAuthorizationHeader(canString, auth), nil
+}
+
+func (c Client) buildCanonicalizedResource(uri string, auth authentication) (string, error) {
+	errMsg := "buildCanonicalizedResource error: %s"
+	u, err := url.Parse(uri)
+	if err != nil {
+		return "", fmt.Errorf(errMsg, err.Error())
+	}
+
+	cr := "/" + c.getCanonicalizedAccountName()
+
+	if len(u.Path) > 0 {
+		// Any portion of the CanonicalizedResource string that is derived from
+		// the resource's URI should be encoded exactly as it is in the URI.
+		// -- https://msdn.microsoft.com/en-gb/library/azure/dd179428.aspx
+		cr += u.EscapedPath()
+	}
+
+	params, err := url.ParseQuery(u.RawQuery)
+	if err != nil {
+		return "", fmt.Errorf(errMsg, err.Error())
+	}
+
+	// See https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Core/Util/AuthenticationUtility.cs#L277
+	if auth == sharedKey {
+		if len(params) > 0 {
+			cr += "\n"
+			keys := make([]string, 0, len(params))
+			for key := range params {
+				keys = append(keys, key)
+			}
+
+			sort.Strings(keys)
+
+			for i, key := range keys {
+				if len(params[key]) > 1 {
+					sort.Strings(params[key])
+				}
+
+				if i == len(keys)-1 {
+					cr += fmt.Sprintf("%s:%s", key, strings.Join(params[key], ","))
+				} else {
+					cr += fmt.Sprintf("%s:%s\n", key, strings.Join(params[key], ","))
+				}
+			}
+		}
+	} else {
+		// search for "comp" parameter, if exists then add it to canonicalizedresource
+		for key := range params {
```

Comment:
Params is type Values which is just an alias for a map. There's no need to do this in a loop, it can be simplified to:

``` go
else if val, isPresent := params["comp"]; isPresent {
    cr += "?comp=" + val[0]
}
```